pkg/scheduler/algorithm/priorities/resource_bin_packing.go

```diff
@@ -0,0 +1,249 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package priorities
+
+import (
+	"fmt"
+	"math"
+	"strings"
+
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	v1helper "k8s.io/kubernetes/pkg/apis/core/v1/helper"
+	"k8s.io/kubernetes/pkg/scheduler/algorithm"
+	priorityutil "k8s.io/kubernetes/pkg/scheduler/algorithm/priorities/util"
+	schedulerapi "k8s.io/kubernetes/pkg/scheduler/api"
+	schedulercache "k8s.io/kubernetes/pkg/scheduler/cache"
+)
+
+// ResourceBinPacking contains information to calculate bin packing priority.
+type ResourceBinPacking struct {
+	resource  v1.ResourceName
+	resources map[v1.ResourceName]int
+}
+
+// NewResourceBinPacking creates a ResourceBinPackingPriorityMap.
+func NewResourceBinPacking(resource v1.ResourceName, resources map[v1.ResourceName]int) (algorithm.PriorityMapFunction, algorithm.PriorityReduceFunction) {
+	resourceBinPackingPrioritizer := &ResourceBinPacking{
+		resource:  resource,
```

Comment:
what's different between `resource` and `resources`?

Reply:
resource is for a single resource of type v1.ResourceName, this does not require assigning weights to the resource for bin packing and should suffice for 90 percent of the users who intend to bin pack a single extended resource.
resources is of type map[v1.ResourceName] int for the special use case where people want to use multiple resources to bin pack and need to define weights for each resource.

Reply:
I'm thinking to use a slice of a struct (resource + weight) for this; the weight is 1 if not set.